R/bm-tpc-h.R

```diff
@@ -0,0 +1,206 @@
+#' Benchmark TPC-H queries
+#'
+#' @section Parameters:
+#' * `engine` One of `c("arrow", "duckdb", "dplyr)`
+#' * `query_num` integer, 1-22
+#' * `format` One of `c("parquet", "feather", "native")`
+#' * `scale` Scale factor to use for data generation (e.g. 0.1, 1, 10, 100)
+#'
+#' @export
+tpc_h <- Benchmark("tpc_h",
+  setup = function(engine = "arrow",
+                   query_num = 1,
+                   format = "parquet",
+                   scale = 10) {
+    # engine defaults to arrow
+    engine <- match.arg(engine, c("arrow", "duckdb", "dplyr"))
+    # input format
+    format <- match.arg(format, c("parquet", "feather", "native"))
+    # query_num defaults to 1 for now
+    stopifnot(
+      "query_num must be an int" = query_num %% 1 == 0,
+      "query_num must 1-22" = query_num >= 1 & query_num <= 22
+    )
+
+    library("dplyr")
+
+    # ensure that we have the _base_ tpc-h files (in parquet)
+    tpch_files <- ensure_source("tpch", scale = scale)
+
+    # these will be filled in later, when we have file paths available
+    input_functions <- list()
+
+    # we use this connection both to populate views/tables and get answer info
+    con <- DBI::dbConnect(duckdb::duckdb())
+
+    if (engine == "arrow") {
+      # ensure that we have the right kind of files available
+      # but for native, make sure we have a feather file, and we will read that
+      # in to memory before the benchmark (below)
+      format_to_convert <- format
+      if (format == "native") {
+        format_to_convert <- "feather"
+      }
+
+      tpch_files <- vapply(
+        tpch_files,
+        ensure_format,
+        FUN.VALUE = character(1),
+        format = format_to_convert
+      )
+
+      # specify readers for each format
+      if (format == "parquet") {
+        input_functions[["arrow"]] <- function(name) {
+          file <- tpch_files[[name]]
+          return(arrow::read_parquet(file, as_data_frame = FALSE))
```

Comment:
Why not open_dataset()? I wonder if that has different/better properties, even if it is a single file. (Hey we should benchmark that!) It at least shouldn't be memory constrained.

Reply:
I'll add that as an option and see if there are better properties. You're right we probably should always do `open_dataset()` here even if these are typically one-file datasets.